keps/sig-cluster-lifecycle/kubeadm/2568-kubeadm-non-root-control-plane/README.md

```diff
@@ -0,0 +1,1022 @@
+  - Run `kube-proxy` as non-root.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+Here is what we propose to do at a high level:
+
+1. Run the control-plane components as non-root, by assigning a unique uid/gid to the containers in the Pods using the `runAsUser` and `runAsGroup` fields in the [securityContext](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#securitycontext-v1-core) on the Pods.
+
+2. Drop all capabilities from kube-controller-manager and kube-scheduler Pods. Drop all but cap_net_bind_service from kube-apiserver Pod, this will allow us to run as non-root while still being able to bind to ports < 1024. This can be achieved by using the `capabilities` field in the [securityContext](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#securitycontext-v1-core) of the containers in the Pods.
+
+3. Set the seccomp profile to `runtime/default`, this can be done by setting the `seccompProfile` field in the `securityContext`.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+As a security conscious user I would like to run the kubernetes control-plane as non-root.
+
+#### Story 2
+
+As a user of kubeadm which sells it self as the best-practice and fast way to bootstrap the cluster, I expect it to follow security best practices like running containers as non-root.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+### How do we assign these uid/gids to the components?
```

Comment:
As per kubeadm office hours discussion, this is probably the trickiest part of this KEP.

Kubeadm strives for simplicity, and ideally we should avoid adding new knobs.

However, picking a constant value for users could prevent users to use Kubeadm, so we have to think how to make this change to land in a non disruptive way, and eventually, how to make it possible to override the default.

A possible approach could be:
- have this feature under a feature gate, so the user have to explicitly opt-in
- use constant values
- document how to override values using kubeadm patches

This would allow to implement the feature, and starting to collect user feedback, so we can plane for a more informed decision about this point while graduating this feature to alpha.
Opinions?

Reply:
Having a feature gate makes sense to me to start. It'd allow us to pick and generate UID/GID values for the user, if they opt to do so. We should still allow users to customize these values though.

Reply:
The approach that you mentioned SGTM.

> thank you for the initial draft @vinayakankugoyal
> 
> i've added a few comments on which we can thread discussion.
> you seem to have skipped a number of sections that should be filled in the KEP.
> 

I plan to add those sections once we have some consensus on what approach we take to assign uid/gids.

> **something that i would like to see in this KEP** is to grab all mounts from `/etc/kubernetes/manifests/*.yaml` and do a breakdown on what we are going to do with the permissions of all of them.
> 
> some of the mounts are secrets / credentials and i would like to see outlined in this KEP what permissions /owners we set on them if they are not owned by 0/0.

Done. I added a table for the files/directories that are important/relevant and what permissions we should set on it.

Reply:
I think we need to follow distro standards here.

For Debian and friends, it is:
https://www.debian.org/doc/debian-policy/ch-opersys.html#uid-and-gid-classes
And RH family:
https://docs.fedoraproject.org/en-US/packaging-guidelines/UsersAndGroups/

These boil down to needing to either execute `adduser --system` or at least pick a free user between SYS_UID_MIN and SYS_UID_MAX and a free group between SYS_GID_MIN and SYSD_GID_MAX in /etc/login.defs.

Reply:
lol I was just about to suggest adduser/groupadd myself. Great idea!

Reply:
can we detect what uid/gid(s) are free and "allocate" such in the +1000 range ideally?
then during kubeadm component creation we inform the users of the values on stdout.

Reply:
^ and that can be the default behavior? 

prior discussions between kubeadm maintainers indicated that we should not allow the user configure the uid/gid per component....but if there is a lot of demand we can add as optional in v1beta3.

Reply:
I am summarizing the approach in this comment. 
1. Feature gate for users to opt-in
2. Instead of hard coding the values in kubeadm we find available uids in the 1000+ range and write the deets to std out. (this is the default behavior if the feature-gate is enabled, but no uid/gids have been supplied)
3. We allow users to set the uid/gid if they wish to via v1beta3.

Does this seem reasonable to all?

Reply:
from my perspective, yes.

@randomvariable 

> These boil down to needing to either execute adduser --system or at least pick a free user between SYS_UID_MIN and SYS_UID_MAX and a free group between SYS_GID_MIN and SYS_GID_MAX in /etc/login.defs.

are you suggesting to use the SYS_ ranges for these components?

these and non-sys ranges can be obtained by parsing:
/etc/login.defs

Reply:
The file is available in different distros. It's documented in `adduser`'s manpage.  This should be used in preference to finding a free UID and adding 1000, as that's likely to land in the user range.

Reply:
ok, i guess we should search for a free UID in the SYS range from /etc/login.defs